### PR TITLE
[Querier] use connection pool

### PR DESCRIPTION
**Pull request by sharang**
## This PR is for:

- Querier

In the test environment, we constructed five kinds of query pressure with QPS ranging from 16 to 56, and each query pressure lasted for 300 seconds. During the pressure test, we found two problems: 1) There will be thousands of connections inside the deepflow-server Pod ClickHouse Socket, and 90% of the status is TIME_WAIT; 2) When the QPS pressure suddenly increases, it will almost certainly lead to write failure, and the reason for the failure is that deepflow-server disconnects the socket written to ClickHouse.

After investigation, it was found that the write failure when QPS changed was caused by Querier using a short connection to query ClickHouse. In addition, a large number of TIME_WAIT sockets were also related to unreasonable MaxOpenConns/MaxIdleConns values.

### Added benchmark

```bash
[root@prometheus-benchmark-test001 xiangyang]# cat query-qps.sh

time_end=$(( `date +%s` ))
(( time_start = time_end - 7200 ))
duration=300

metrics=demo_cpu_usage_seconds_total

while :; do
for qps in 20 40 60 80; do
        echo "`date` - Target QPS $qps"
        cmd="wrk2 -c 10 -t 10 -L -R $qps -d $duration \"http://10.96.0.117:20416/prom/api/v1/query_range?start=${time_start}&end=${time_end}&step=60s&query=avg_over_time(rate(${metrics}[1m])[2m:10s])\" | grep -E \"^( 50\\.| 99\\.000|Requests/sec)\""
        echo $cmd
        eval $cmd
        echo
        sleep 1
done
done
```

#### before

- Writes often fail due to socket reset:
```
SELECT
    event_time,
    query_duration_ms,
    exception_code,
    exception
FROM query_log
WHERE (type = 'ExceptionWhileProcessing') AND ((databases[1]) = 'ext_metrics') AND (query_kind = 'Insert')
ORDER BY event_time DESC
LIMIT 10

Query id: ec3e5f0d-d618-415b-86e8-24db339a5a10

┌──────────event_time─┬─query_duration_ms─┬─exception_code─┬─exception────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ 2023-03-12 20:17:39 │               966 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while reading from socket (100.101.95.66:47918). (NETWORK_ERROR) (version 22.8.6.71 (official build)) │
│ 2023-03-12 20:08:12 │                 2 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while writing to socket (100.101.95.66:60690). (NETWORK_ERROR) (version 22.8.6.71 (official build))   │
│ 2023-03-12 20:08:12 │               803 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while reading from socket (100.101.95.66:60692). (NETWORK_ERROR) (version 22.8.6.71 (official build)) │
│ 2023-03-12 19:37:10 │                 1 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while writing to socket (100.101.95.66:45464). (NETWORK_ERROR) (version 22.8.6.71 (official build))   │
│ 2023-03-12 19:25:15 │               401 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while reading from socket (100.101.95.66:58010). (NETWORK_ERROR) (version 22.8.6.71 (official build)) │
│ 2023-03-12 19:25:14 │                 1 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while writing to socket (100.101.95.66:58012). (NETWORK_ERROR) (version 22.8.6.71 (official build))   │
└─────────────────────┴───────────────────┴────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
┌──────────event_time─┬─query_duration_ms─┬─exception_code─┬─exception────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ 2023-03-12 12:59:13 │                 2 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while writing to socket (100.101.95.67:58518). (NETWORK_ERROR) (version 22.8.6.71 (official build))   │
│ 2023-03-12 12:44:21 │                 1 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while writing to socket (100.101.95.67:40428). (NETWORK_ERROR) (version 22.8.6.71 (official build))   │
│ 2023-03-12 12:35:31 │               403 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while reading from socket (100.101.95.67:50470). (NETWORK_ERROR) (version 22.8.6.71 (official build)) │
│ 2023-03-12 12:16:35 │                 1 │            210 │ Code: 210. DB::NetException: Connection reset by peer, while writing to socket (100.101.95.68:49606). (NETWORK_ERROR) (version 22.8.6.71 (official build))   │
└─────────────────────┴───────────────────┴────────────────┴──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```
- thousands of TIME_WAIT sockets
```
[root@prometheus-benchmark-test001 xiangyang]# kubectl exec -it -n deepflow deepflow-server-765c879b9-wzmh7 -- sh
/ # /bin/netstat -antp | grep 9000 | wc -l
3702
/ # /bin/netstat -antp | grep 9000 | grep -v TIME_WAIT | wc -l
36
```

#### after

- Write failures no longer occur even though QPS continues to increase to 80/s.
- The number of sockets is reduced to around 100, and there are almost no TIME_WAIT issues.
```
[root@prometheus-benchmark-test001 xiangyang]# kubectl exec -it -n deepflow deepflow-server-75ff78cf8b-zlbmp -- sh
/ # /bin/netstat -antp | grep 9000 | wc -l
116
/ # /bin/netstat -antp | grep 9000 | grep -v TIME_WAIT | wc -l
115
```

### CODE REVIEW

File: server/querier/engine/clickhouse/clickhouse.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ func (e *CHEngine) ExecuteQuery(args *common.QuerierParams) (*common.Result, map
 	var sqlList []string
 	var err error
 	sql := args.Sql
-	query_uuid := args.QueryUUID
+	query_uuid := args.QueryUUID // FIXME: should be queryUUID
```
Comment (sharang) [inline server/querier/engine/clickhouse/clickhouse.go:61]: 应该用驼峰风格

File: server/querier/engine/clickhouse/common/utils.go [hunk 2]
```diff
@@ -203,10 +203,10 @@ func GetExtTables(db string, ctx context.Context) (values []interface{}) {
 	}
 	sql := ""
 	if db == "ext_metrics" {
-		sql = "select table from ext_metrics_custom_field group by table"
+		sql = "SELECT table FROM flow_tag.ext_metrics_custom_field GROUP BY table"
```
Comment (sharang) [inline server/querier/engine/clickhouse/common/utils.go:206]: 关键词大写

File: server/querier/engine/clickhouse/metrics/ext_common.go [hunk 3]
```diff
@@ -45,7 +45,7 @@ func GetExtMetrics(db, table, where string, ctx context.Context) (map[string]*Me
 		var externalMetricSql string
 		var tableFilter string
 		var whereSql string
-		externalMetricSql = "SELECT field_name, table FROM %s_custom_field WHERE %s field_type='metrics' %s GROUP BY field_name, table ORDER BY table, field_name ASC"
+		externalMetricSql = "SELECT field_name, table FROM flow_tag.%s_custom_field WHERE %s field_type='metrics' %s GROUP BY field_name, table ORDER BY table, field_name ASC"
```
Comment (sharang) [inline server/querier/engine/clickhouse/metrics/ext_common.go:48]: 显式写明 db。

但目前存在一个 db 写了两次的问题，要排查，出错现象是当查询 ext_metrics 的 show tag values 时，最终发给 clickhouse 的语句为：SELECT .... FROM flow_tag.`flow_tag.xxxxx_custom_field_value`

@SongZhen0704 @xiaochaoren1

File: server/querier/engine/clickhouse/client/client.go [hunk 4]
```diff
@@ -62,18 +66,38 @@ func (c *Client) init(query_uuid string) error {
 			IP:        c.Host,
 		}
 	}
-	url := fmt.Sprintf("clickhouse://%s:%s@%s:%d/%s?&query_id=%s", c.UserName, c.Password, c.Host, c.Port, c.DB, query_uuid)
-	conn, err := splunksqlx.Open("clickhouse", url)
-	if err != nil {
-		log.Errorf("connect clickhouse failed: %s, url: %s, query_uuid: %s", err, url, query_uuid)
-		return err
+	if connection == nil { // FIXME: RWLock
+		conn, err := clickhouse.Open(&clickhouse.Options{
```
Comment (sharang) [inline server/querier/engine/clickhouse/client/client.go:70]: 这里做了两件事：
1. 将短连接修改为了长连接，这样修改后看到的现象是写入失败不再有了
   - 否则只要打 QPS 压测，16/26/36/46/56 五种 QPS 下都会有写入失败，特别是 QPS 变化的几秒钟内几乎必定写入失败
2. 修改了不合理的 MaxOpenConns 和 MaxIdleConns，默认值还是会导致有大量 TIME_WAIT sockets

还剩余两个 TODO @SongZhen0704 @xiaochaoren1 :
1. max-clickhouse-connection 需要加入 server.yaml
2. 这样修改后去掉了追踪的能力，后面应当补回来

File: server/querier/prometheus/converters.go [hunk 5]
```diff
@@ -108,7 +108,7 @@ func PromReaderTransToSQL(req *prompb.ReadRequest) (sql string, db string, datas
 	if db == "" || db == chCommon.DB_NAME_EXT_METRICS || db == chCommon.DB_NAME_DEEPFLOW_SYSTEM {
 		metrics = append(metrics, extMetricsTagsName)
 	} else {
-		showSql := fmt.Sprintf("SHOW tags FROM %s WHERE time >= %d AND time <= %d", table, startTime, endTime)
+		showSql := fmt.Sprintf("SHOW tags FROM %s.%s WHERE time >= %d AND time <= %d", db, table, startTime, endTime)
```
Comment (sharang) [inline server/querier/prometheus/converters.go:111]: 上面说的错误的语句可能是这里产生的：

```
SELECT .... FROM flow_tag.`flow_tag.xxxxx_custom_field_value` ...
```

File: server/querier/statsd/statsd.go [hunk 6]
```diff
@@ -3,7 +3,7 @@ package statsd
 import (
 	"sync"
 
-	"github.com/deepflowio/deepflow/server/libs/stats"
+	"github.com/deepflowio/deepflow/server/libs/stats" // FIXME: why not use stats directly
```
Comment (sharang) [inline server/querier/statsd/statsd.go:6]: 这个文件将所有 query 的监控都放在这里了，这是不合理的，不同的子模块可能有不同的 tag 需求、指标需求，监控是和业务高度关联的，把所有子模块业务流程的监控写到一个地方不合理。  @SongZhen0704 @xiaochaoren1 后面应当修改。

File: server/querier/engine/clickhouse/client/client.go [hunk 7]
```diff
@@ -98,38 +120,37 @@ func (c *Client) DoQuery(params *QueryParams) (result *common.Result, err error)
...
+	columns := rows.ColumnTypes()
 	resColumns := len(columns)
 	if err != nil {
 		c.Debug.Error = fmt.Sprintf("%s", err)
 		return nil, err
 	}
 	columnNames := make([]interface{}, 0, len(columns))
-	columnTypes := make([]string, 0, len(columns))
-	var columnSchemas common.ColumnSchemas
+	var columnSchemas common.ColumnSchemas // FIXME: Slice growth should be avoided.
```
Comment (sharang) [inline server/querier/engine/clickhouse/client/client.go:130]: 这里的 slice 需要避免 growth。定义时就要确定它的 cap @SongZhen0704 @xiaochaoren1

File: server/querier/engine/clickhouse/client/format.go [hunk 8]
```diff
@@ -34,37 +36,44 @@ var VALUE_TYPE_MAP = map[string]int{
...
+		return int(*v), VALUE_TYPE_INT, nil
+	case *uint64:
+		return int(*v), VALUE_TYPE_INT, nil
+	case *uint32:
+		return int(*v), VALUE_TYPE_INT, nil
+	case *time.Time:
+		return v.String(), VALUE_TYPE_STRING, nil
+	case *net.IP:
+		return v.String(), VALUE_TYPE_STRING, nil
+	case **float64: // FIXME: Why does this type exist?
```
Comment (sharang) [inline server/querier/engine/clickhouse/client/format.go:61]: @SongZhen0704 @xiaochaoren1 为何会有这种类型的存在？而不是 *float64

File: server/querier/prometheus/converters.go [hunk 9]
```diff
@@ -271,7 +271,7 @@ func RespTransToProm(result *common.Result) (resp *prompb.ReadResponse, err erro
 		} else if metricsType == "Float64" {
 			metricsValue = values[metricsIndex].(float64)
 		} else {
-			metricsValue = *values[metricsIndex].(*float64)
+			return nil, errors.New(fmt.Sprintf("Unknown metrics type %s, value = %v", metricsType, values[metricsIndex]))
```
Comment (sharang) [inline server/querier/prometheus/converters.go:274]: 原来这里的处理是非常粗糙的，这里应当报错才对，在 format.go 中覆盖 **float64 的情况。

File: server/querier/engine/clickhouse/tag/description.go [hunk 10]
```diff
@@ -521,7 +521,8 @@ func GetTagValues(db, table, sql string) (*common.Result, []string, error) {
 	if strings.Contains(strings.ToLower(sql), "like") || strings.Contains(strings.ToLower(sql), "regexp") {
 		orderBy = "length(display_name)"
 	}
-	sql = fmt.Sprintf("SELECT value,name AS display_name, description FROM flow_tag.%s WHERE tag_name='%s' %s GROUP BY value, display_name, description ORDER BY %s ASC %s", table, tag, whereSql, orderBy, limitSql)
+	// querier will be called later, so there is no need to display the declaration db
+	sql = fmt.Sprintf("SELECT value,name AS display_name, description FROM %s WHERE tag_name='%s' %s GROUP BY value, display_name, description ORDER BY %s ASC %s", table, tag, whereSql, orderBy, limitSql)
```
Comment (sharang) [inline server/querier/engine/clickhouse/tag/description.go:525]: value, 后面加个空格

半角标点后面都要形成加空格的习惯

File: server/server.yaml [hunk 11]
```diff
@@ -193,6 +193,7 @@ querier:
     host: clickhouse
     port: 9000
     timeout: 60
+    max-clickhouse-connection: 20
```
Comment (sharang) [inline server/server.yaml:196]: 由于是放在 clickhouse 下面的，直接就叫  max-connection 吧 @xiaochaoren1

**Review by xiaochaoren1 — DISMISSED** [review]

**Review by xiaochaoren1 — APPROVED** [review]

**Pull request merged**